discussion/rust-for-embedded-systems.tex

```diff
@@ -0,0 +1,90 @@
+
+\section{Rust for Embedded Systems} % (fold)
+\label{sec:disc:rust_for_arm}
+% \todo{Mention this in discussion. Is there a better way of solving this? Currently it fetches a release tarball of the language. It fetches more than it needs to for each library, but it's quicker/easier than using e.g. git to do it. It does however take a little while to compile all the dependencies on smaller machines.}
+
+Before we started on this thesis, we were already aware that there were a couple of other projects that were running {\rust} bare-metal on ARM \glspl{mcu}.
+We did, however, not know how easy it would be to get up and running with {\rust} on the EFM32.
+It took us less than one day to implement and execute a {\rust} blink-demo on the {\STK}.
+The {\rust} documentation was sufficient for figuring out how to define non-std {\rust} programs (i.e. the ones that are annotated with \texttt{\#[no\_std]}), and projects like \texttt{armboot} helped us build the project for the ARM architecture.
+{\rust}'s runtime is comparable to {\C}'s, because the startup routine in the two languages is the same, but {\rust}'s {\core} library is larger and provides more functionality than the {\C} standard library, which we consider to be an added bonus of {\rust}.
+
+\subsection{The Standard Library} % (fold)
+\label{sub:using_the_standard_library}
+
+% subsection using_the_standard_library (end)
+
+{\rust} has a rich standard library that offers a strong foundation for all the programs that want to depend on it.
+As we described in \autoref{ssub:rust:features}, the {\std} crate acts like an abstraction over all the different libraries in \gls{rsl} and the {\rustc} compiler is itself dependent on a couple of traits to be present in this module.
+It is easy enough to define programs that are not dependent on this {\std} library, and instead only dependent on {\core}, but we think the difference between {\std} and {\core} is too large.
+
+\gls{rcl} does provide us with enough functionality to write idiomatic {\rust} code, but we had to implement a few workarounds in order to get support for {\rust}'s standard constructs for dynamic memory allocation.
+Currently, {\rust} does not have functionality to provide, or automatically compile, any of its standard libraries for target architectures other than the ones it supports by default.
+Because our project targets an ARM processor architecture, we have to conditionally download and compile the libraries (like \lib{alloc} and \lib{collections}) that have functionality that we want to utilize, as described in \autoref{ssub:transitioning_to_cargo}.
+This is not necessarily a big problem in itself, as it does not complicate the process much more than the need to explicitly define the libraries as custom dependencies in the \texttt{Cargo.toml}.
+We do, however, think that the distinction between standard and non-standard programs make it considerably harder to enable us to use the great ecosystem that {\rust} is surrounded by.
+
```

Comment:
Append:

`Handling these dependencies before the Cargo process was introduced was a huge pain point. The inclusion of the Cargo enabled us to easly stay up to date with the nightly releases of the compiler.`